app/src/main/java/ch/epfl/cs311/wanderwave/model/remote/BeaconConnection.kt

```diff
@@ -0,0 +1,111 @@
+package ch.epfl.cs311.wanderwave.model.remote
+
+import android.util.Log
+import ch.epfl.cs311.wanderwave.model.data.Beacon
+import ch.epfl.cs311.wanderwave.model.data.Location
+import ch.epfl.cs311.wanderwave.model.data.Track
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.filterNotNull
+
+class BeaconConnection : FirebaseConnectionInt<Beacon, Beacon> {
+
+  override val collectionName: String = "beacons"
+
+  override val getItemId = { beacon: Beacon -> beacon.id }
+
+  val trackConnection = TrackConnection()
+
+  private val db = FirebaseFirestore.getInstance()
+
+  // Document to Beacon
+  override fun documentToItem(document: DocumentSnapshot): Beacon {
+    val id = document.id
+    val locationMap = document.get("location") as? Map<String, Any>
+    val latitude = locationMap?.get("latitude") as? Double ?: 0.0
+    val longitude = locationMap?.get("longitude") as? Double ?: 0.0
+    val name = locationMap?.get("name") as? String ?: ""
+    val location = Location(latitude, longitude, name)
+
+    val tracks = listOf<Track>()
+
+    return Beacon(id = id, location = location, tracks = tracks)
+  }
+
+  override fun addItem(item: Beacon) {
+    super.addItem(item)
+    trackConnection.addItemsIfNotExist(item.tracks)
+  }
+
+  override fun addItemWithId(item: Beacon) {
+    super.addItemWithId(item)
+    trackConnection.addItemsIfNotExist(item.tracks)
+  }
+
+  override fun updateItem(item: Beacon) {
+    super.updateItem(item)
+    trackConnection.addItemsIfNotExist(item.tracks)
+  }
+
+  override fun getItem(itemId: String): Flow<Beacon> {
+    val dataFlow = MutableStateFlow<Beacon?>(null)
+
+    db.collection(collectionName)
+        .document(itemId)
+        .get()
+        .addOnSuccessListener { document ->
+          Log.d("Firestore", "Document: $document")
+          Log.d("Firestore", "test : ${document.data}")
+          if (document != null && document.data != null) {
+            val beacon = documentToItem(document)
+            Log.d("Firestore", "Beacon: $beacon")
+            val trackRefs = document.get("tracks") as? List<DocumentReference>
+            val tracks = mutableListOf<Track>()
+            trackRefs?.forEach { trackRef ->

```

Comment:
Maybe try to modularize the trackRefs treatment, because it's quite long for a lambda in my opinion

Reply:
Will try to do that thanks.